observableeditablebuffer.go

```diff
@@ -0,0 +1,194 @@
+package main
+
+import (
+	"fmt"
+	"io"
+)
+
+// The ObservableEditableBuffer is used by the main program
+// to add, remove and check on the current observer(s) for a Text.
+// Text in turn, implements BufferObserver for the various required callback functions in BufferObserver.
+type ObservableEditableBuffer struct {
+	currobserver BufferObserver
+	observers    map[BufferObserver]struct{} // [private I think]
+	f            *File
+}
+
+// AddObserver adds e as an observer for edits to this File.
+func (e *ObservableEditableBuffer) AddObserver(observer BufferObserver) {
+	if e.observers == nil {
+		e.observers = make(map[BufferObserver]struct{})
+	}
+	e.observers[observer] = struct{}{}
+	e.currobserver = observer
+
+}
+
+// DelObserver removes e as an observer for edits to this File.
+func (e *ObservableEditableBuffer) DelObserver(observer BufferObserver) error {
+	if _, exists := e.observers[observer]; exists {
+		delete(e.observers, observer)
+		if observer == e.currobserver {
+			for k := range e.observers {
+				e.currobserver = k
+				break
+			}
+		}
+		return nil
+	}
+	return fmt.Errorf("can't find editor in File.DelObserver")
+}
+
+// SetCurObserver sets the current observer.
+func (e *ObservableEditableBuffer) SetCurObserver(observer BufferObserver) {
+	e.currobserver = observer
+}
+
+// GetCurObserver gets the current observer and returns it as a interface type.
+func (e *ObservableEditableBuffer) GetCurObserver() interface{} {
+	return e.currobserver
+}
+
+// AllObservers preforms tf(all observers...)
+func (e *ObservableEditableBuffer) AllObservers(tf func(i interface{})) {
+	for t := range e.observers {
+		tf(t)
+	}
+}
+
+// GetObserverSize will return the size of the observer map.
+func (e *ObservableEditableBuffer) GetObserverSize() int {
+	return len(e.observers)
+}
+
+// HasMultipleObservers returns true if their are multiple observers to the File.
+func (e *ObservableEditableBuffer) HasMultipleObservers() bool {
+	return len(e.observers) > 1
+}
+
+// MakeObservableEditableBuffer is a constructor wrapper for NewFile() to abstract File from the main program.
+func MakeObservableEditableBuffer(filename string, b RuneArray) *ObservableEditableBuffer {
+	f := NewFile(filename)
+	f.b = b
+	oeb := &ObservableEditableBuffer{
+		currobserver: nil,
+		observers:    nil,
+		f:            f,
+	}
+	oeb.f.oeb = oeb
+	return oeb
+}
+
+// MakeObservableEditableBufferTag is a constructor wrapper for NewTagFile() to abstract File from the main program.
+func MakeObservableEditableBufferTag(b RuneArray) *ObservableEditableBuffer {
+	f := NewTagFile()
+	f.b = b
+	oeb := &ObservableEditableBuffer{
+		currobserver: nil,
+		observers:    nil,
+		f:            f,
+	}
+	oeb.f.oeb = oeb
+	return oeb
+}
+
+// Clean is a forwarding function for file.Clean.
+func (e *ObservableEditableBuffer) Clean() {
+	e.f.Clean()
+}
+
+// FileSize is a forwarding function for file.Size.
+func (e *ObservableEditableBuffer) FileSize() int {
```

Comment:
I'd call this `Size()`. Remember that it will eventually be in package `file` so if named this, people would write `file.FileSize()`. `file.Size()` seems nicer.